### PR TITLE
The new 3.0 version from the CloudFest hackathon

### PR DESCRIPTION
**Pull request by 2ndkauboy**

### CODE REVIEW

File: inc/Fields/Honeypot.php [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+	$options['field_name'];
+	 * @param $markup
+	 * @param $options
+	 */
+	public static function inject( $markup, $options ) {
+		$dom = new DOMDocument();
+		$dom->loadHTML( $markup, LIBXML_HTML_NOIMPLIED | LIBXML_HTML_NODEFDTD );
+		$xpath = new DOMXPath( $dom );
+		$input = $xpath->query( '//*[@id="' . $options['field_id'] . '"]' )->item( 0 );
+		if ( ! $input ) {
```
Comment (2ndkauboy) [inline inc/Fields/Honeypot.php:25]: @angelocali Are we sure `ext-dom` is available by default on hosts? :thinking: If not, but we will need it, we should add a check to `antispam_bee.php` so we make sure the plugin is not loaded if the extension is missing.

File: inc/Fields/Honeypot.php [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+					$output .= ' name="' . esc_attr( self::get_secret_name_for_post() ) . '" ';
+					$output .= $matches['between1'] . $matches['between2'] . $matches['between3'];
+					$output .= $matches['after'] . '>';
+					$output .= $matches['content'];
+					$output .= '</textarea><textarea ' . $attributes_string . '></textarea>';
+
+					$output .= $id_script;
+					return $output;
+				}, $markup );
```
[the remark this answers is not in the document]
Comment (angelocali) [reply inc/Fields/Honeypot.php]: @2ndkauboy Yes, it’s a first approach to achieve this goal. :)

File: inc/Fields/Honeypot.php (hunk 1 above)
Comment (angelocali) [reply inc/Fields/Honeypot.php]: @2ndkauboy From the [WordPress Handbook](https://make.wordpress.org/hosting/handbook/server-environment/#php-extensions) I see that `ext-dom` is strongly recommended, but WordPress also works without it. Would you rather deactivate the plugin, rule or try an alternative approach without DOMDocument if the extension isn’t available?

File: inc/Fields/Honeypot.php [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+							(?P<before3>[^>]*)
+							name=["\']{{HONEYPOT_NAME}}["\']
+							(?P<between3>[^>]*)
+						)
+						(?P<after>[^>]*)
+						>
+						(?s)(?P<content>.*?)
+						<\/textarea>
+					)/x'
+				);
```
[the remark this answers is not in the document]
Comment (florianbrinkmann) [reply inc/Fields/Honeypot.php]: We replaced that to make it compatible with textareas that have other names and ids, apart from that the regex should be the same

**Comment by angelocali** [comment]
I looked into the code for the settings page and I suggest to add a class named `Tab` as a container for sections. I feel more comfortable if a section doesn’t automatically create a new tab. We could also have various filters to extend tabs and tab-sections. What do you think?

File: lib/Admin/SettingsPage.php [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+			}
+		}
+	}
+
+	/**
+	 * Settings page content.
+	 */
+	public function options_page() {
+		?>
+		<div class="wrap" id="ab_main">
```
Comment (pfefferle) [inline lib/Admin/SettingsPage.php:94]: I would use templates for this, so we do not have to `echo` too much HTML

File: antispam_bee.php [hunk 5]
```diff
@@ -1,3041 +1,88 @@
...
-	 * @since  2.10.0 Changes needed because of new way to add the honeypot field via filter instead of output buffering
-	 *
-	 * @param string $data Markup of the comment field or whole page (depending on ob option).
-	 */
-	public static function prepare_comment_field( $data ) {
-		if ( empty( $data ) ) {
-			return $data;
-		}
+	// Check, if the min. required PHP version is available and if not, show an admin notice.
+	if ( version_compare( PHP_VERSION, '7.2', '<' ) ) {
```
[the remark this answers is not in the document]
Comment (2ndkauboy) [reply antispam_bee.php]: Was fixed in 490e94c4232ef6d919b258257bdaf220dcd30eda

File: inc/Fields/Honeypot.php (hunk 1 above)
Comment (2ndkauboy) [reply inc/Fields/Honeypot.php]: I've checked with PHP7.4 and `ext-dom` is not included in `php7.4-common`, so I've added another check in 5d189b24cc616a3855e38c3a3490c09566eed1f5

File: lib/Rules/InitRule.php [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+trait InitRule {
+	public static function init() {
+		add_filter( 'asb_rules', function ( $rules ) {
+			$rules[] = [
+				'verifiable' => self::class,
+			];
+
+			return $rules;
+		} );
```
[the remark this answers is not in the document]
Comment (2ndkauboy) [reply lib/Rules/InitRule.php]: This has been handled in e3e24d42126ad22c8ad2702bbb4bc0f774d4cc40

File: lib/Admin/SettingsPage.php (hunk 4 above)
Comment (2ndkauboy) [reply lib/Admin/SettingsPage.php:94]: How would a template solve this, having less `echos` calls Can you give an example?

File: lib/load.php [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+			call_user_func( [ $module, 'init' ] );
+		}
+	}
+
+	add_filter(
+		'comment_form_field_comment',
+		function ( $field_markup ) {
+			return HoneypotField::inject( $field_markup, [ 'field_id' => 'comment' ] );
+		}
+	);
```
[the remark this answers is not in the document]
Comment (florianbrinkmann) [reply lib/load.php:83]: Yes, that would be better. We added it there to be able to test if the Honeypot integration works without much overhead

**Comment by florianbrinkmann** [comment]
To discuss: Should we somehow combine the rules _Delete spam_ and _Delete spam by reason_ ? Because if both are dedicated options, and both are checked, one overwrites the other and leads to maybe unwanted behavior like all spam is deleted instead of only for a few reasons

File: lib/Rules/IsActive.php [hunk 8]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace AntispamBee\Rules;
+
+use AntispamBee\Helpers\Settings;
+
+trait IsActive {
+	public static function is_active( $type ) {
+		return true;
```
[the remark this answers is not in the document]
Comment (florianbrinkmann) [reply lib/Rules/IsActive.php:9]: The branch is work in progress :)

**Pull request closed without merging**